pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotJoinExchangeNodeInsertRule.java

```diff
@@ -52,22 +54,132 @@ public void onMatch(RelOptRuleCall call) {
     RelNode left = PinotRuleUtils.unboxRel(join.getInput(0));
     RelNode right = PinotRuleUtils.unboxRel(join.getInput(1));
     JoinInfo joinInfo = join.analyzeCondition();
+    Map<String, String> joinHintOptions = PinotHintOptions.JoinHintOptions.getJoinHintOptions(join);
+    PinotHintOptions.DistributionType leftDistributionType;
+    PinotHintOptions.DistributionType rightDistributionType;
+    if (joinHintOptions != null) {
+      leftDistributionType = PinotHintOptions.JoinHintOptions.getLeftDistributionType(joinHintOptions);
+      rightDistributionType = PinotHintOptions.JoinHintOptions.getRightDistributionType(joinHintOptions);
+    } else {
+      leftDistributionType = null;
+      rightDistributionType = null;
+    }
     RelNode newLeft;
     RelNode newRight;
     if (PinotHintOptions.JoinHintOptions.useLookupJoinStrategy(join)) {
-      // Lookup join - add local exchange on the left side
-      newLeft = PinotLogicalExchange.create(left, RelDistributions.SINGLETON);
+      // Lookup join
```

Comment:
(nit) this function is somewhat less readable. we should consider refactoring

we should also add validation. for example, a query can't have a exchange hint for left side as BROADCAST and right side as HASH for the same JOIN op.

Reply:
Refactored and extracted the common code.
I added a TODO for the validation for hash join because I haven't thought through which combination is invalid. User might decide to use a certain distribution based on their data layout, and I don't want to fail them prematurely.